### PR TITLE
Do not attempt to re-hover on exiting subplots

### PR DESCRIPTION
**Pull request by etpinard**
hopefully fixes https://github.com/plotly/plotly.js/issues/4157 and related `Cannot read _subplot of undefined` issues.

@plotly/plotly_js this bug has been plaguing dash users for some times now (see https://community.plot.ly/t/dash-app-with-js-clientside-callback-sometimes-crashes/29599/5 for a reproducible case).

### CODE REVIEW

**Comment by etpinard** [comment]
To include in `v1.50.1`

File: src/plots/cartesian/graph_interact.js [hunk 1]
```diff
@@ -59,7 +59,7 @@ exports.initInteractions = function initInteractions(gd) {
                 // This is on `gd._fullLayout`, *not* fullLayout because the reference
                 // changes by the time this is called again.
                 gd._fullLayout._rehover = function() {
-                    if(gd._fullLayout._hoversubplot === subplot) {
+                    if((gd._fullLayout._hoversubplot === subplot) && gd._fullLayout._plots[subplot]) {
```
Comment (archmoj) [inline src/plots/cartesian/graph_interact.js]: What about putting this condition test inside `Fx.hover`?

Comment (etpinard) [reply src/plots/cartesian/graph_interact.js]: Why would that be better?

Comment (archmoj) [reply src/plots/cartesian/graph_interact.js]: Just in case that the function is called from different places.
Otherwise; this performs better.

**Comment by alexcjohnson** [comment]
Nice 🎉 presumably this works for disappearing non-cartesian subplots as well as cartesian?

**Comment by etpinard** [comment]
> presumably this works for disappearing non-cartesian subplots as well as cartesian?

Presumably, we shouldn't be calling `Plots.rehover` on non-cartesian subplots ... but hold on looks like mapbox subplots define a `_rehover` function too:

image

Thanks for the tip!

I'll double-check if something similar happens to mapbox subplots before merging this PR.

File: src/plots/cartesian/graph_interact.js (hunk 1 above)
Comment (etpinard) [reply src/plots/cartesian/graph_interact.js]: You're making a good point here, consider:

```js
        /*var data = [
            {y: [1], hoverinfo: 'y'},
            {y: [2], hoverinfo: 'y', xaxis: 'x2', yaxis: 'y2'}
        ];*/

        var data2 = [
            {y: [1], hoverinfo: 'y'}
        ];

        var layout = {
            xaxis: {domain: [0, 0.5]},
            xaxis2: {anchor: 'y2', domain: [0.5, 1]},
            yaxis2: {anchor: 'x2'},
            width: 400,
            height: 200,
            margin: {l: 0, t: 0, r: 0, b: 0},
            showlegend: false
        };
Plotly.newPlot(gd, data2, layout)

// which is part of our semi-official API
Plots.Fx.hover(gd, {}, 'x2y2')
```

fails with the same `Cannot read _subplot of undefined ` error.

---

So, I'll keep the check and `graph_interact.js` (to not have to call `Fx.hover` for nothing) AND  handle the empty subplot case in `Fx.hover` (+ a test).

Thanks for the headsup @archmoj !

**Comment by alexcjohnson** [comment]
> Presumably, we shouldn't be calling `Plots.rehover` on non-cartesian subplots

that's not the implication of these comments... anyway regardless of the eventual *fix*, the *test* is pretty clear, hover on a subplot of each type while it disappears. Seems a bit overkill for a very simple fix though, also I've never seen users complaining about this with anything but cartesian. But it would be nice to have at least manually tested that this works.

https://github.com/plotly/plotly.js/blob/25fe1c929672461e0b915a737134f6a975edf8f8/src/components/fx/hover.js#L227-L241

File: src/plots/cartesian/graph_interact.js (hunk 1 above)
Comment (etpinard) [reply src/plots/cartesian/graph_interact.js]: done in https://github.com/plotly/plotly.js/pull/4269/commits/89e52f71eb663bbaa0e045757c78c38454991569

**Comment by etpinard** [comment]
@archmoj this PR is ready for a 2nd round of review.

File: src/components/fx/hover.js [hunk 2]
```diff
@@ -224,24 +224,20 @@ function _hover(gd, evt, subplot, noHoverEvent) {
...
-
-            // TODO make sure that fullLayout_plots axis refs
-            // get updated properly so that we don't have
-            // to use Axes.getFromId in general.
-
-            xaArray[i] = Axes.getFromId(gd, plotObj.xaxis._id);
-            yaArray[i] = Axes.getFromId(gd, plotObj.yaxis._id);
-            continue;
+            xaArray[i] = plots[spId].xaxis;
+            yaArray[i] = plots[spId].yaxis;
```
Comment (etpinard) [inline src/components/fx/hover.js]: `gd._fullLayout._plots[/* subplot id */].(x|y).axis` is now always filled in properly thanks for [`Plots.linkSubplots`](https://github.com/plotly/plotly.js/blob/25fe1c929672461e0b915a737134f6a975edf8f8/src/plots/plots.js#L825-L946) called during `Plots.supplyDefaults`.

The `TODO` comment above was over three years old.

**Comment by archmoj** [comment]
Excellent.
:dancer: :dancer:

**Pull request merged**